### PR TITLE
gotify: refactor, use nix-update

### PR DESCRIPTION
**Pull request by yu-re-ka**
## Description of changes

- move version, source-sha to gotify-server expression
- gotify-server.ui: rewrite with stdenv.mkDerivation, inherit version, src from gotify-server
- inherit ui yarn offlineCache into gotify-server.passthru, so that nix-update-script can find it
- use nix-update-script

nix-update-script supports updating the goModules vendorSha256
After https://github.com/Mic92/nix-update/pull/68 is merged, it will also update the yarn offlineCache hash.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/gotify/ui.nix [hunk 1]
```diff
@@ -1,67 +1,47 @@
...
+    yarn install --offline --frozen-lockfile --ignore-scripts --no-progress --non-interactive
+    patchShebangs node_modules/
+  '';
 
   buildPhase = ''
-    export HOME=$(mktemp -d)
-    export WRITABLE_NODE_MODULES="$(pwd)/tmp"
-    export NODE_OPTIONS=--openssl-legacy-provider
-    mkdir -p "$WRITABLE_NODE_MODULES"
+    runHook preConfigure
```
Comment (doronbehar) [inline pkgs/servers/gotify/ui.nix:32]:
```suggestion
    runHook preBuild
```

File: pkgs/servers/gotify/ui.nix [hunk 2]
```diff
@@ -1,67 +1,47 @@
...
+    export NODE_OPTIONS=--openssl-legacy-provider
+    yarn --offline build
 
-    # In `node_modules/.bin` are relative symlinks that would be broken after copying them over,
-    # so we take care of them here.
-    mkdir -p "$WRITABLE_NODE_MODULES"/.bin
-    for x in "$node_modules"/.bin/*; do
-      ln -sfv "$node_modules"/.bin/"$(readlink "$x")" "$WRITABLE_NODE_MODULES"/.bin/"$(basename "$x")"
-    done
+    runHook postConfigure
```
Comment (doronbehar) [inline pkgs/servers/gotify/ui.nix:37]:
```suggestion
    runHook postBuild
```

File: pkgs/servers/gotify/ui.nix [hunk 3]
```diff
@@ -1,67 +1,47 @@
...
+  '';
 
-    ln -sfv "$WRITABLE_NODE_MODULES" node_modules
-    cd ../..
+  installPhase = ''
+    runHook preConfigure
```
Comment (doronbehar) [inline pkgs/servers/gotify/ui.nix:41]:
```suggestion
    runHook preInstall
```

File: pkgs/servers/gotify/ui.nix [hunk 4]
```diff
@@ -1,67 +1,47 @@
...
 
-    yarn build
+    mv build $out
 
-    cd deps/gotify-ui
-    rm -rf node_modules
-    ln -sf $node_modules node_modules
-    cd ../..
+    runHook postConfigure
```
Comment (doronbehar) [inline pkgs/servers/gotify/ui.nix:45]:
```suggestion
    runHook postInstall
```

**Review by doronbehar — COMMENTED** [review]
Wow amazing, looks great! Thanks

File: pkgs/servers/gotify/default.nix [hunk 5]
```diff
@@ -5,19 +5,20 @@
...
   # `update.sh` script.
-  version = import ./version.nix;
+  version = "2.2.4";
 
   src = fetchFromGitHub {
     owner = "gotify";
     repo = "server";
     rev = "v${version}";
-    sha256 = import ./source-sha.nix;
+    sha256 = "1kc4l95hrhi7lb9x8gy19xpwj12j4syg6w1kbllf3g3k83sr444f";
```
Comment (dotlambda) [inline pkgs/servers/gotify/default.nix:21]:
```suggestion
    hash = "1kc4l95hrhi7lb9x8gy19xpwj12j4syg6w1kbllf3g3k83sr444f";
```

File: pkgs/servers/gotify/default.nix [hunk 6]
```diff
@@ -26,7 +27,7 @@ buildGoModule rec {
   #   server[780]: stat /var/lib/private/ui/build/index.html: no such file or directory
   allowGoReference = true;
 
-  vendorSha256 = import ./vendor-sha.nix;
+  vendorSha256 = "sha256-TxxiyfWzlzQ2R2hgeBzB11FIiOz5rIBfaIm15DQ+dL0=";
```
Comment (dotlambda) [inline pkgs/servers/gotify/default.nix:30]:
```suggestion
  vendorHash = "sha256-TxxiyfWzlzQ2R2hgeBzB11FIiOz5rIBfaIm15DQ+dL0=";
```

File: pkgs/servers/gotify/default.nix [hunk 7]
```diff
@@ -5,19 +5,20 @@
...
   # `update.sh` script.
-  version = import ./version.nix;
+  version = "2.2.4";
 
   src = fetchFromGitHub {
     owner = "gotify";
     repo = "server";
     rev = "v${version}";
-    sha256 = import ./source-sha.nix;
+    hash = "1kc4l95hrhi7lb9x8gy19xpwj12j4syg6w1kbllf3g3k83sr444f";
```
Comment (dotlambda) [inline pkgs/servers/gotify/default.nix:21]: This should be turned into an SRI hash.

**Review by dotlambda — COMMENTED** [review]
Can you squash?

**Comment by yu-re-ka** [comment]
Done

File: pkgs/servers/gotify/default.nix [hunk 8]
```diff
@@ -35,11 +36,14 @@ buildGoModule rec {
   ui = callPackage ./ui.nix { };
 
   preBuild = ''
-    cp -r ${ui}/libexec/gotify-ui/deps/gotify-ui/build ui/build && go run hack/packr/packr.go
+    if [ -n "$ui" ]; then
```
Comment (doronbehar) [inline pkgs/servers/gotify/default.nix:39]: Why did you add this conditional?

**Comment by doronbehar** [comment]
Thanks again for all the work @yu-re-ka . I promise that question will be the last iteration!

File: pkgs/servers/gotify/default.nix (hunk 8 above)
Comment (yu-re-ka) [reply pkgs/servers/gotify/default.nix]: I added a comment:

> to make the preBuild a no-op inside the goModules fixed-output derivation, where it would fail

Comment (doronbehar) [reply pkgs/servers/gotify/default.nix]: > Why did you add this conditional?

I would have never have imagined. Thanks for that comment!

**Comment by doronbehar** [comment]
Will merge when CI is green.

**Comment by doronbehar** [comment]
@yu-re-ka something is wrong with `gotify-server.ui`, I get:

```
gotify-server> unpacking sources
gotify-server> unpacking source archive /nix/store/mv491kvyrgscxz17xzn8d8vjk3qhmjbf-source
gotify-server> source root is source
gotify-server> patching sources
gotify-server> updateAutotoolsGnuConfigScriptsPhase
gotify-server> configuring
gotify-server> building
gotify-server> cp: cannot stat '/nix/store/q3snajjfb1snzp8zyjab746zk65nl58j-gotify-ui-2.2.4/libexec/gotify-ui/deps/gotify-ui/build': No such file or directory
gotify-server> /nix/store/7cni7ndy2pm18ysl5znq6znb30sxp156-stdenv-linux/setup: line 87: pop_var_context: head of shell_variables not a function context
gotify-server> /nix/store/7cni7ndy2pm18ysl5znq6znb30sxp156-stdenv-linux/setup: line 1596: pop_var_context: head of shell_variables not a function context
```

**Comment by doronbehar** [comment]
I think I know how to fix this, not too hard.

**Comment by yu-re-ka** [comment]
I swear I had made this change already, just like the missing comment about the conditional. must have gotten lost somewhere

**Comment by doronbehar** [comment]
Builds now. Now I'm testing the update via:

```
nix run github:Mic92/nix-update -- gotify-server
```

**Comment by doronbehar** [comment]
And it fails with:

```
building

syntax: prefetch-yarn-deps [path to yarn.lock] [options]

Options:
  -h --help         Show this help
  -v --verbose      Verbose output
  --builder         Only perform the download to current directory, then exit
error: builder for '/nix/store/q36yma457qyhv9xhlsxangdrjisc9baw-offline.drv' failed with exit code 1;
       last 10 log lines:
       > configuring
       > no configure script, doing nothing
       > building
       >
       > syntax: prefetch-yarn-deps [path to yarn.lock] [options]
       >
       > Options:
       >   -h --help         Show this help
       >   -v --verbose      Verbose output
       >   --builder         Only perform the download to current directory, then exit
       For full logs, run 'nix log /nix/store/q36yma457qyhv9xhlsxangdrjisc9baw-offline.drv'.
Traceback (most recent call last):
  File "/nix/store/iayfpkqy1ngfiys9sixd8mf8q1raizcf-nix-update/bin/.nix-update-wrapped", line 9, in <module>
    sys.exit(main())
             ^^^^^^
  File "/nix/store/iayfpkqy1ngfiys9sixd8mf8q1raizcf-nix-update/lib/python3.11/site-packages/nix_update/__init__.py", line 287, in main
    package = update(options)
              ^^^^^^^^^^^^^^^
  File "/nix/store/iayfpkqy1ngfiys9sixd8mf8q1raizcf-nix-update/lib/python3.11/site-packages/nix_update/update.py", line 371, in update
    update_yarn_deps_hash(opts, package.filename, package.yarn_deps)
  File "/nix/store/iayfpkqy1ngfiys9sixd8mf8q1raizcf-nix-update/lib/python3.11/site-packages/nix_update/update.py", line 261, in update_yarn_deps_hash
    target_hash = nix_prefetch(opts, "offlineCache")
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/iayfpkqy1ngfiys9sixd8mf8q1raizcf-nix-update/lib/python3.11/site-packages/nix_update/update.py", line 126, in nix_prefetch
    raise UpdateError(
nix_update.errors.UpdateError: failed to retrieve hash when trying to update gotify-server.offlineCache
```

**Comment by doronbehar** [comment]
There is an update available now, so testing the patch to `nix-update` should be convenient!

**Comment by yu-re-ka** [comment]
try again ;)

**Comment by doronbehar** [comment]
Damn that was delicate!

Checked that `nix run github:Mic92#nix-update -- gotify-server` works, and pushed that update commit. I noticed that the `offlineCache` hash didn't update, and indeed, upstream's `ui/yarn.lock` file didn't change in a year. I hope the observation that the `offlineCache` hash shouldn't change in this update is correct. I will merge when CI is green.

**Comment by doronbehar** [comment]
Thanks again for all the work! It was impresive.

**Comment by doronbehar** [comment]
Weird... ofborg's builds fail on all platforms. What do you think about its failure @yu-re-ka ? Maybe could you run a `nixpkgs-review` on this? Although I think thought that some of the fixed output derivations are available in your cache...

**Comment by yu-re-ka** [comment]
The new error is not an artifact and caused by the update ;)

**Comment by doronbehar** [comment]
> The new error is not an artifact and caused by the update ;)

Right right right, I haven't tested the update actually. Adapting to it was easy thankfully :).
